src/ray/core_worker/core_worker.cc

```diff
@@ -933,17 +933,21 @@ void CoreWorker::RegisterToGcs() {
 }
 
 void CoreWorker::CheckForRayletFailure() {
+  bool should_shutdown = false;
   // When running worker process in container, the worker parent process is not raylet.
   // So we add RAY_RAYLET_PID enviroment to ray worker process.
   if (auto env_pid = RayConfig::instance().RAYLET_PID(); !env_pid.empty()) {
     auto pid = static_cast<pid_t>(std::stoi(env_pid));
     if (!IsProcessAlive(pid)) {
       RAY_LOG(ERROR) << "Raylet failed. Shutting down. Raylet PID: " << pid;
-      Shutdown();
+      should_shutdown = true;
     }
   } else if (!IsParentProcessAlive()) {
     RAY_LOG(ERROR) << "Raylet failed. Shutting down.";
-    Shutdown();
+    should_shutdown = true;
+  }
+  if (should_shutdown) {
+    task_execution_service_.post([this]() { Shutdown(); }, "CoreWorker.Shutdown");
```

Comment:
Can you asserts / comments on the Shutdown handler on why it has to be called from that thread?

Reply:
+1, it will be great if we can somehow assert in Shutdown() and other functions that they indeed run in the specific thread.

Reply:
+1 on @mwtian maybe we can consider tracking thread id of the core worker and assert inside the `Shutdown` method?